src/cider/exec/nextgen/operators/HashJoinNode.cpp

```diff
@@ -0,0 +1,253 @@
+              .ret_type = JITTypeTag::BOOL,
+              .params_vector = {{fixsize_val.getValue().get(), index_.get()}}});
+      return row_data;
+    } else {
+      JITTypeTag tag = utils::getJITTypeTag(expr_->get_type_info().get_type());
+      // data buffer decoder
+      auto data_pointer = fixsize_val.getValue()->castPointerSubType(tag);
+      auto row_data = data_pointer[index_];
+      return row_data;
+    }
+  }
+
+ private:
+  utils::JITExprValue& buffer_values_;
+  ExprPtr& expr_;
+  JITValuePointer& index_;
+};
+
+TranslatorPtr HashJoinNode::toTranslator(const TranslatorPtr& succ) {
+  return createOpTranslator<HashJoinTranslator>(shared_from_this(), succ);
+}
+
+void HashJoinTranslator::consume(context::CodegenContext& context) {
+  codegen(context);
+}
+
+// traverse join_quals expr tree to get join key value and null vector
+void traverse(ExprPtr expr,
+              jitlib::JITFunctionPointer func,
+              std::vector<JITValuePointer>& keys,
+              std::vector<JITValuePointer>& nulls) {
+  if (Analyzer::ColumnVar* col_var = dynamic_cast<Analyzer::ColumnVar*>(expr.get())) {
+    if (col_var->get_table_id() == 100) {
+      utils::FixSizeJITExprValue jit_value(col_var->codegen(*func));
+      keys.emplace_back(jit_value.getValue());
+      // null vector handle
+      if (!expr->get_type_info().get_notnull()) {
+        nulls.emplace_back(jit_value.getNull());
+      }
+    }
+  } else {
+    auto children = expr->get_children_reference();
+    for (auto child : children) {
+      traverse(*child, func, keys, nulls);
+    }
+  }
+}
+
+void HashJoinTranslator::codegen(context::CodegenContext& context) {
+  auto func = context.getJITFunction();
+  auto join_quals = dynamic_cast<HashJoinNode*>(node_.get())->getJoinQuals();
+
+  std::vector<JITValuePointer> keys;
+  std::vector<JITValuePointer> nulls;
+  for (int i = 0; i < join_quals.size(); ++i) {
+    traverse(join_quals[i], func, keys, nulls);
+  }
+
+  // open up a section of buffer to reserve the join result
+  auto join_res_buffer = context.registerBuffer(
+      16, "join_res_buffer", [](context::Buffer* buf) {}, false);
+
+  // pack join key values(support only one key now)
+  auto key_value = func->packJITValues<64>(keys);
+  // pack null
+  auto key_null = func->packJITValues<8>(nulls);
+  // register hashtable
+  auto hashtable = context.registerHashTable();
+
+  // TODO(qiuyang) : hashtable will be a base class pointer
+  auto join_res_len = func->emitRuntimeFunctionCall(
+      "look_up_value_by_key",
+      JITFunctionEmitDescriptor{
+          .ret_type = JITTypeTag::INT64,
+          .params_vector = {
+              hashtable.get(), key_value.get(), key_null.get(), join_res_buffer.get()}});
+
+  auto build_table_map = dynamic_cast<HashJoinNode*>(node_.get())->getBuildTableMap();
+  auto row_index = func->createVariable(JITTypeTag::INT64, "row_index", 0l);
+  row_index = func->createLiteral(JITTypeTag::INT64, 0l);
```

Comment:
`row_index` should can be created as `Literal` directly.

Reply:
I'm affraid we can't. I use it as variable in the logic behind , otherwise it will call "JITValue is not a variable in LLVMJITValue::assign." error

Reply:
@bigPYJ1151 is this expected?

Reply:
Yes, only two types of ```JITValue``` could be assigned:
- variable.
- dereference of ```Pointer``` type ```JITValue```.